packages/design-system-scripts/gulp/lint.js

```diff
@@ -0,0 +1,94 @@
+const eslint = require('gulp-eslint');
+const eslintConfig = require('../eslint.config');
+const gulp = require('gulp');
+const gulpIf = require('gulp-if');
+const path = require('path');
+const stylelint = require('gulp-stylelint');
+const stylelintConfig = require('../stylelint.config');
+const getPackageName = require('./common/getPackageName');
+const streamPromise = require('./common/streamPromise');
+const { log, logTask } = require('./common/logUtil');
+const { CORE_SOURCE_PACKAGE } = require('./common/constants');
+
+/**
+ * Class naming pattern
+ * ~~~~~~~~~~~~~~~~~~~~
+ * Names:    ds-
+ * Prefixes: l- c- u-
+ * Pattern:  [NAMESPACE]-[PREFIX]-[BLOCK]__[ELEMENT]--[MODIFIER]
+ * Examples: .ds-c-button--primary, .ds-c-card__title, .ds-u-text-underlined
+ */
+const coreSassClassNamingPattern = /^(ds-)(l|c|u|)(-[a-z0-9]+)((--?|__)[a-z0-9]+)*$/;
+
+// Lint Sass files using stylelint. Further configuration for CSS linting
+// can be handled in stylelint.config.js
+async function lintSass(dir, fix) {
+  const src = path.join(dir, 'src');
+  const config = _.cloneDeep(stylelintConfig);
+
+  // Add class naming pattern linting for core DS
+  const name = await getPackageName(dir);
+  if (name === CORE_SOURCE_PACKAGE) {
+    _.set(config, ['rules', 'selector-class-pattern'], coreSassClassNamingPattern);
+  }
+
+  return streamPromise(
+    gulp
+      .src([`${src}/**/*.scss`, `!${src}/**/*.docs.scss}`])
+      .pipe(changedInPlace({ firstPass: true }))
+      .pipe(count(`## Sass files linted in ${src}`))
+      .pipe(
+        stylelint({
+          config,
+          fix,
+          failAfterError: process.env.NODE_ENV === 'test',
+          reporters: [{ formatter: 'string', console: true }],
+          syntax: 'scss'
+        })
+      )
+      .pipe(gulpIf(fix, gulp.dest(src)))
+  );
+}
+
+// Lint JS files using eslint. Further configuration for JS linting
+// can be handled in eslint.config.js
+async function lintJS(dir, fix) {
+  // Taken from gulp-eslint example
+  // https://github.com/adametry/gulp-eslint/blob/master/example/fix.js
+  const isFixed = (file) => {
+    return file.eslint != null && file.eslint.fixed;
+  }
+
+  const src = path.join(dir, 'src');
+  return streamPromise(
+    gulp
+      .src([`${src}/**/*.{js,jsx}`, `!${src}/**/*.test.{js,jsx}`])
+      .pipe(count(`## JS files linted in ${src}`))
+      .pipe(eslint({ ...{ fix }, ...eslintConfig }))
+		  .pipe(eslint.result(result => {
+        // eslint.format() is not working some reason
+        // Manually output the results instead
+        const msg = stylish([result])
+        if (msg) {
+          log(msg)
+        }
+      }))
+      .pipe(gulpIf(isFixed, gulp.dest(src)))
+      .pipe(gulpIf(process.env.NODE_ENV === 'test', eslint.failAfterError()))
+  );
+}
+
+module.exports = {
+  async lintDirectories(directories, fix) {
+    logTask('🔎 ', `Linting "src" directory in: ${directories.join(', ')}`);
+    // TODO: Fix failAfterError with streams
```

Comment:
@pwolfert Our previous scripts configured linting to `failAfterError` when the env = test. I'm guessing this is mainly so that Travis CI will error when linting fails. 

When I test `failAfterError` on this branch, the linting task just stops for the directory its currently working on, and the task just completes early. I think this might have to do with the way we are promisifying streams. I was referring to [this](https://nodejs.org/api/stream.html#stream_stream_finished_stream_options_callback) but I havent been able to find a solution yet.

Reply:
I just played around with this for a while, and I couldn't get it to throw an error either, and then I reread that documentation you linked to like the fourth time and realized that `error: false` is what we want here
```js
function streamPromise(stream) {
  return finished(stream, { readable: false, writeable: false, error: false });
}
```
At first I was thinking the default _true_ value is what we wanted.

Reply:
Changing that line in `packages/design-system-scripts/gulp/common/streamPromise.js` made it throw the error and fail for me

Reply:
Ahh thank you!